more_modbus/ModbusReader.cpp

```diff
@@ -378,4 +386,112 @@ void ModbusReader::readDevice(const std::shared_ptr<ModbusDevice>& device)
+                                                     rewritable->getBoolValue()))
+                        {
+                            LOG(DEBUG) << "Successfully rewrote '" << rewritable->getReference() << "' - " << diff
+                                       << "ms.";
+                            rewritable->update(rewritable->getBoolValue());
+                            ++succeededMappings;
+                        }
+                        else
+                        {
+                            LOG(DEBUG) << "Failed to rewrite '" << rewritable->getReference() << "'.";
+                        }
+                    }
+                    else
+                    {
+                        if (rewritable->getRegisterCount() == 1)
+                        {
+                            if (m_modbusClient.writeHoldingRegister(rewritable->getSlaveAddress(),
+                                                                    rewritable->getStartingAddress(),
+                                                                    rewritable->getBytesValues().front()))
+                            {
+                                LOG(DEBUG)
+                                  << "Successfully rewrote '" << rewritable->getReference() << "' - " << diff << "ms.";
+                                rewritable->update(rewritable->getBytesValues());
+                                ++succeededMappings;
+                            }
+                            else
+                            {
+                                LOG(DEBUG) << "Failed to rewrite '" << rewritable->getReference() << "'.";
+                            }
+                        }
+                        else
+                        {
+                            if (m_modbusClient.writeHoldingRegisters(
+                                  rewritable->getSlaveAddress(), rewritable->getStartingAddress(),
+                                  const_cast<std::vector<std::uint16_t>&>(rewritable->getBytesValues())))
+                            {
+                                LOG(DEBUG)
+                                  << "Successfully rewrote '" << rewritable->getReference() << "' - " << diff << "ms.";
+                                rewritable->update(rewritable->getBytesValues());
+                                ++succeededMappings;
+                            }
+                            else
+                            {
+                                LOG(DEBUG) << "Failed to rewrite '" << rewritable->getReference() << "'.";
+                            }
+                        }
+                    }
+                }
+            }
+
+            // If all the groups had error while reading, report the device as having errors.
+            if (requiredMappings > 0 && succeededMappings == 0)
+            {
+                m_deviceActiveStatus[device->getSlaveAddress()] = false;
```

Comment:
This is modified in read thread also, so it is definitely a data race. It is also edited in run thread with no synchronization mechanism whatsoever.
This also is runs every millisecond while run thread checks status every m_readPeriond + 100ms. I think this entire read/write/reconnect needs a rework